### PR TITLE
docs(blog-post): replicate spatial dev guru blog

### PR DESCRIPTION
**Pull request by ncclementi**
## Description of changes
Replicate Spatial Dev Guru's blog using Ibis

## Issues closed

Closes https://github.com/ibis-project/ibis/issues/7717

### CODE REVIEW

File: docs/posts/ibis-duckdb-geospatial-dev-guru/index.qmd [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+## Let's get started
+
+Because this dataset does not contain any geometries, we have to load the spatial extension. If the dataset included any geometry columns, Ibis is smart enough to load the extension for us upon reading the data.
+
+```{python}
+con = ibis.duckdb.connect("biketrip.ddb")
+con.load_extension("spatial")
+
+#read data and rename columns to use snake case
```
Comment (cpcloud) [inline docs/posts/ibis-duckdb-geospatial-dev-guru/index.qmd:60]:
```suggestion
# read data and rename columns to use snake case
```

File: docs/posts/ibis-duckdb-geospatial-dev-guru/index.qmd [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+biketrip = con.read_parquet("202003-citibike-tripdata.parquet").rename("snake_case")
+biketrip
+```
+
+
```
Comment (cpcloud) [inline docs/posts/ibis-duckdb-geospatial-dev-guru/index.qmd:65]:
```suggestion
```

**Comment by ncclementi** [comment]
🤔 : I see on the preview this extra black bar, not sure why. 

Screenshot 2024-01-12 at 12 55 57 PM

File: docs/posts/ibis-duckdb-geospatial-dev-guru/index.qmd [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+We have the information about the longitude and latitude for start and end stations,
+to create geometry points and put the spatial features to use.
+
+## Create bike trip table
+
+In the original tutorial, Spatial Dev Guru creates a table with transformed
+"Pickup" and "Dropoff" points. In DuckDB the `st_transform` function takes by default
+points as `YX` (latitude/longitude) while in Ibis,  we assume data in the form
```
Comment (cpcloud) [inline docs/posts/ibis-duckdb-geospatial-dev-guru/index.qmd:73]:
```suggestion
points as `YX` (latitude/longitude) while in Ibis, we assume data in the form
```

File: docs/posts/ibis-duckdb-geospatial-dev-guru/index.qmd [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+`XY` (longitude/latitude) to be consistent with PostGIS and Geopandas.
+
+In Ibis the function to tr
```
Comment (cpcloud) [inline docs/posts/ibis-duckdb-geospatial-dev-guru/index.qmd:76]: Missing some text?

File: docs/posts/ibis-duckdb-geospatial-dev-guru/index.qmd [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+# Notice longitude/latitude order
+pickup = biketrip.start_station_longitude.point(biketrip.start_station_latitude).name("pickup")
+dropoff = biketrip.end_station_longitude.point(biketrip.end_station_latitude).name("dropoff")
+
+#convert is the equivalent of `st_transform`
+biketrip = biketrip.mutate(pickup_point=pickup.convert('EPSG:4326', 'EPSG:3857'),
+                           dropoff_point=dropoff.convert('EPSG:4326', 'EPSG:3857')
+                        )
+
+biketrip[["pickup_point", "dropoff_point"]]
```
Comment (cpcloud) [inline docs/posts/ibis-duckdb-geospatial-dev-guru/index.qmd:87]: These line are pretty long. Can you use `_` to condense them a bit? Also should be formatted.

File: docs/posts/ibis-duckdb-geospatial-dev-guru/index.qmd [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+spatial points for pickup and dropoff locations, that are in the Web Mercator projection ([EPSG:3857](https://epsg.io/3857)).
+
+## Identify popular starts and end stations
+
+The following queries retrieve a list of bike start and end stations with their respective trip count in descending order.
+
+**Top 10 start stations by trip count**
+
+```{python}
+biketrip.group_by(biketrip.start_station_name).agg(trips = ibis._.count()).order_by(ibis.desc("trips"))
```
Comment (cpcloud) [inline docs/posts/ibis-duckdb-geospatial-dev-guru/index.qmd:100]:
```suggestion
biketrip.start_station_name.topk(10, by=_.trips.count())
```

Does that produce the same result?

File: docs/posts/ibis-duckdb-geospatial-dev-guru/index.qmd [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+```
+
+**Top 10 end stations by trip count**
+
+```{python}
+biketrip.group_by(biketrip.end_station_name).agg(trips = ibis._.count()).order_by(ibis.desc("trips"))
```
Comment (cpcloud) [inline docs/posts/ibis-duckdb-geospatial-dev-guru/index.qmd:106]:
```suggestion
biketrip.end_station_name.topk(10, by=_.trips.count())
```

Might work, similar to the `start_station_name` variant.

File: docs/posts/ibis-duckdb-geospatial-dev-guru/index.qmd [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+We can also calculate the average trip duration and distance traveled for each
+user type. According to the [data dictionary](https://ride.citibikenyc.com/system-data), user type can be "customer" or "subscriber" where:
+
+- Customer = 24-hour pass or 3-day pass user
+- Subscriber = Annual Member
+
+```{python}
+biketrip.group_by(_.usertype).aggregate(
+    avg_duration=biketrip.tripduration.mean(),
+    avg_distance=biketrip.pickup_point.distance(biketrip.dropoff_point).name("trip_distance").mean()
```
Comment (cpcloud) [inline docs/posts/ibis-duckdb-geospatial-dev-guru/index.qmd:120]:
```suggestion
    avg_distance=biketrip.pickup_point.distance(biketrip.dropoff_point).mean()
```

No reason to call `.name` in between other expressions.

File: docs/posts/ibis-duckdb-geospatial-dev-guru/index.qmd [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+```
+
+## Analyzing efficiency: trip duration vs linear distance
+
+The original tutorial defines `efficiency_ratio` as the `trip_duration` / `linear_distance`, where a higher efficiency ratio could mean a more direct
+route or faster travel times.
+
+```{python}
+# linear distance
+trip_distance = biketrip.pickup_point.distance(biketrip.dropoff_point).name("trip_distance")
```
Comment (cpcloud) [inline docs/posts/ibis-duckdb-geospatial-dev-guru/index.qmd:131]:
```suggestion
trip_distance = biketrip.pickup_point.distance(biketrip.dropoff_point)
```

No need to name this.

File: docs/posts/ibis-duckdb-geospatial-dev-guru/index.qmd [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+
+biketrip = biketrip.mutate(linear_distance = trip_distance, efficiency_ratio = biketrip.tripduration / trip_distance)
```
Comment (cpcloud) [inline docs/posts/ibis-duckdb-geospatial-dev-guru/index.qmd:133]: Can you format this line?

File: docs/posts/ibis-duckdb-geospatial-dev-guru/index.qmd [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+The following query shows all the bike trips whose pickup point falls within a
+500 meter radius of the first point of the table with `long=-74.00552427` and
+`lat=40.71146364`.
+
+```{python}
+# grab the first row of the data
+first_point = biketrip.limit(1)
+
+trips_within_500 = biketrip.filter(_.pickup_point.within(
+    first_point.select(_.pickup_point.buffer(500)).to_array()))
```
Comment (cpcloud) [inline docs/posts/ibis-duckdb-geospatial-dev-guru/index.qmd:159]: Can you format this line?

**Review by cpcloud — CHANGES_REQUESTED** [review]
Looking good! Mostly stylistic comments!

**Comment by cpcloud** [comment]
> 🤔 : I see on the preview this extra black bar, not sure why.

I think this is an artifact of DuckDB's progress bar.

You can turn it off with 

```sql
SET enable_progress_bar = true;
```

File: docs/posts/ibis-duckdb-geospatial-dev-guru/index.qmd (hunk 4 above)
Comment (ncclementi) [reply docs/posts/ibis-duckdb-geospatial-dev-guru/index.qmd:73]: whoops, I'll fix that.

**Comment by deepyaman** [comment]
Minor comment--before it get's merged, can you make sure it's "spatial" instead of "sapatial". :)

File: docs/posts/ibis-duckdb-geospatial-dev-guru/index.qmd (hunk 7 above)
Comment (ncclementi) [reply docs/posts/ibis-duckdb-geospatial-dev-guru/index.qmd]: nice, I wasn't aware of the `topk()` function, the equivalent query would be
`biketrip.end_station_name.topk(10)` In the first case I left both options to showcase the similarity.

**Comment by ncclementi** [comment]
@deepyaman great catch!

**Comment by ncclementi** [comment]
@cpcloud I think I cover all your comments/requests.  About the progress bar, I wasn't able to figure out where to set the variable `SET enable_progress_bar = false`?

**Comment by cpcloud** [comment]
Ah, sorry, that's code to execute with `con.raw_sql`

**Review by cpcloud — APPROVED** [review]

**Comment by cpcloud** [comment]
@ncclementi I can merge this now if you want to follow up with a PR to disable the progress bar, or happy to wait as well.

**Comment by cpcloud** [comment]
SGTM!

**Comment by gforsyth** [comment]
SHIP IT

**Pull request merged**